src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs

```diff
@@ -0,0 +1,249 @@
+        public object GetOrCreateObjectForComInstance(IntPtr externalComObject, CreateObjectFlags flags)
+        {
+            return GetOrCreateObjectForComInstanceEx(externalComObject, flags, null);
+        }
+
+        [DllImport(RuntimeHelpers.QCall)]
+        private static extern void GetOrCreateObjectForComInstanceInternal(ObjectHandleOnStack comWrappersImpl, IntPtr externalComObject, CreateObjectFlags flags, ObjectHandleOnStack wrapper, ObjectHandleOnStack retValue);
+
+        /// <summary>
+        /// Create a managed object for the object pointed at by <paramref name="externalComObject"/> respecting the values of <paramref name="flags"/>.
+        /// </summary>
+        /// <param name="externalComObject">Object to import for usage into the .NET runtime.</param>
+        /// <param name="flags">Flags used to describe the external object.</param>
+        /// <returns>Returns a managed object associated with the supplied external COM object.</returns>
+        /// <remarks>
+        /// If the object cannot be created and <code>null</code> is returned, the call to <see cref="ComWrappers.GetOrCreateObjectForComInstance(IntPtr, CreateObjectFlags)"/> will throw a <see cref="System.ArgumentNullException"/>.
+        /// </remarks>
+        protected abstract object? CreateObject(IntPtr externalComObject, CreateObjectFlags flags);
+
+        // Call to execute the abstract instance function
+        internal static object? CallCreateObject(ComWrappers? comWrappersImpl, IntPtr externalComObject, CreateObjectFlags flags)
+            => (comWrappersImpl ?? s_globalInstance!).CreateObject(externalComObject, flags);
+
+        /// <summary>
+        /// Get the currently registered managed object or uses the supplied managed object and registers it.
+        /// </summary>
+        /// <param name="externalComObject">Object to import for usage into the .NET runtime.</param>
+        /// <param name="flags">Flags used to describe the external object.</param>
+        /// <param name="wrapper">The <see cref="object"/> to be used as the wrapper for the external object</param>
+        /// <returns>Returns a managed object associated with the supplied external COM object.</returns>
+        /// <remarks>
+        /// If the <paramref name="wrapper"/> instance already has an associated external object a <see cref="System.NotSupportedException"/> will be thrown.
+        /// </remarks>
+        public object GetOrRegisterObjectForComInstance(IntPtr externalComObject, CreateObjectFlags flags, object wrapper)
+        {
+            return GetOrCreateObjectForComInstanceEx(externalComObject, flags, wrapper);
+        }
+
+        private object GetOrCreateObjectForComInstanceEx(IntPtr externalComObject, CreateObjectFlags flags, object? wrapper)
+        {
+            if (externalComObject == IntPtr.Zero)
+                throw new ArgumentNullException(nameof(externalComObject));
+
+            ComWrappers impl = this;
```

Comment:
I expect that this will be pretty hot method and that we are going to make two leaner dedicated paths for this once we put it under profiler...

Reply:
Related: It would be interesting to see how fast is this vs. regular built-in COM interop.

Reply:
What is the "leaner" path? There is a single if check.

Reply:
You are passing the extra argument around, etc. It adds overhead. I am just speculating about the perf problems. Actually measuring the perf and comparing it with the build-in would be the best. I hope that this can match it.

Reply:
Okay. That will be done during the feedback phase with CsWinRT.

Reply:
I did run this under a profiler though and this was never even close to being hot. The mere act of allocating was more expensive.

Reply:
I expect that the common path through this is going to be lookup of existing object (ie no allocation).

> the feedback phase with CsWinRT.

You do not need to wait for that. You can just make a few microbenchmarks to compare built-in COM interop with this.

Reply:
>You do not need to wait for that. You can just make a few microbenchmarks to compare built-in COM interop with this.

Since there is close to no overlap between the two activation/creation scenarios I am going to assume the microbenchmarks in question would relate to re-querying the system for the associated RCW/CCW?. The equivalent to these in the built-in system would be [`Marshal.GetIUnknownForObject()`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.marshal.getiunknownforobject) and [`Marshal.GetObjectForIUnknown()`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.marshal.getobjectforiunknown) - are those the APIs you were thinking about for comparison benchmarking?

Reply:
Yep

Reply:
@jkotas Here are the basic results from the two different APIs. I ran this in the opposite order with the same results. The `COMPlus_TieredCompilation` option had no noticeable impact.

Iteration count: 1000000

```
GetOrCreateComInterfaceForObject: - 35.8286 ns/iter
GetIUnknownForObject: - 54.6582 ns/iter
GetOrCreateComInterfaceForObject: - 37.142599999999995 ns/iter
GetIUnknownForObject: - 55.4512 ns/iter
```

```
GetOrCreateObjectForComInstance: - 64.417 ns/iter
GetObjectForIUnknown: - 86.2446 ns/iter
GetOrCreateObjectForComInstance: - 67.9225 ns/iter
GetObjectForIUnknown: - 85.1364 ns/iter
```

Reply:
The numbers for `GetOrCreateComInterfaceForObject` look suspiciously good. What is the lifetime contract for `GetOrCreateComInterfaceForObject` ? Does it return AddRefed IUnknown like GetIUnknownForObject?

Reply:
> The numbers for GetOrCreateComInterfaceForObject look suspiciously good

Yes they do. I am looking at this in a profiler and attempting to ascertain why.

>Does it return AddRefed IUnknown like GetIUnknownForObject?

Yes.

Reply:
From the looks of the trace the issue here is the scenario I chose hit a degenerate case where the system must do a lot more work to query the associated IClassX tables and an aggregation check. This results in at least two more AddRef()/Release() cycles. Factoring that out, the results become almost identical. The new approach is still slightly faster, but that seems to be because the new code to get at the `IUnknown` is just simpler/more direct.

Reply:
Ok, nice!